pkg/internal/request/span.go

```diff
@@ -63,36 +112,184 @@ type PidInfo struct {
 // REMINDER: any attribute here must be also added to the functions SpanOTELGetters,
 // SpanPromGetters and getDefinitions in pkg/internal/export/metric/definitions.go
 type Span struct {
-	Type           EventType
-	IgnoreSpan     IgnoreMode
-	Method         string
-	Path           string
-	Route          string
-	Peer           string
-	PeerPort       int
-	Host           string
-	HostPort       int
-	Status         int
-	ContentLength  int64
-	RequestStart   int64
-	Start          int64
-	End            int64
-	ServiceID      svc.ID // TODO: rename to Service or ResourceAttrs
-	TraceID        trace2.TraceID
-	SpanID         trace2.SpanID
-	ParentSpanID   trace2.SpanID
-	Flags          uint8
-	Pid            PidInfo
-	PeerName       string
-	HostName       string
-	OtherNamespace string
-	Statement      string
+	Type           EventType      `json:"type"`
+	IgnoreSpan     IgnoreMode     `json:"ignoreSpan"`
+	Method         string         `json:"-"`
+	Path           string         `json:"-"`
+	Route          string         `json:"-"`
+	Peer           string         `json:"peer"`
+	PeerPort       int            `json:"peerPort,string"`
+	Host           string         `json:"host"`
+	HostPort       int            `json:"hostPort,string"`
+	Status         int            `json:"status,string"`
+	ContentLength  int64          `json:"-"`
+	RequestStart   int64          `json:"-"`
+	Start          int64          `json:"-"`
+	End            int64          `json:"-"`
+	ServiceID      svc.ID         `json:"-"` // TODO: rename to Service or ResourceAttrs
+	TraceID        trace2.TraceID `json:"traceID"`
+	SpanID         trace2.SpanID  `json:"spanID"`
+	ParentSpanID   trace2.SpanID  `json:"parentSpanID"`
+	Flags          uint8          `json:"flags,string"`
+	Pid            PidInfo        `json:"-"`
+	PeerName       string         `json:"peerName"`
+	HostName       string         `json:"hostName"`
+	OtherNamespace string         `json:"-"`
+	Statement      string         `json:"-"`
 }
 
 func (s *Span) Inside(parent *Span) bool {
 	return s.RequestStart >= parent.RequestStart && s.End <= parent.End
 }
 
+const (
+	kClient   = "CLIENT"
+	kServer   = "SERVER"
+	kProducer = "PRODUCER"
+	kConsumer = "CONSUMER"
+	kInternal = "INTERNAL"
+)
+
+func kindString(span *Span) string {
+	switch span.Type {
+	case EventTypeHTTP, EventTypeGRPC, EventTypeKafkaServer, EventTypeRedisServer:
+		return kServer
+	case EventTypeHTTPClient, EventTypeGRPCClient, EventTypeSQLClient, EventTypeRedisClient:
+		return kClient
+	case EventTypeKafkaClient:
+		switch span.Method {
+		case MessagingPublish:
+			return kProducer
+		case MessagingProcess:
+			return kConsumer
+		}
+	}
+	return kInternal
+}
+
+// helper attribute functions used by JSON serialization
+type SpanAttributes map[string]string
+
+// nolint
+func spanAttributes(s *Span) SpanAttributes {
+	attr := SpanAttributes{}
+
+	// method
```

Comment:
nit: these coments are not necessary IMO as the switches are descriptive enough

Reply:
Sounds good! I will rework this part of the code as per feedback from @grcevski